internal/speakerlist/speakerlist.go

```diff
@@ -0,0 +1,301 @@
+	mconfig := memberlist.DefaultLANConfig()
+	// mconfig.Name MUST be spec.nodeName, as we will match it against Enpoints nodeName in usableNodes()
+	mconfig.Name = nodeName
+	mconfig.BindAddr = bindAddr
+	if bindPort != "" {
+		mlport, err := strconv.Atoi(bindPort)
+		if err != nil {
+			logger.Log("op", "startup", "error", "unable to parse ml-bindport", "msg", err)
+			return nil, err
+		}
+		mconfig.BindPort = mlport
+		mconfig.AdvertisePort = mlport
+	}
+	loggerout := gokitlog.NewStdlibAdapter(gokitlog.With(sl.l, "component", "Memberlist"))
+	mconfig.Logger = golog.New(loggerout, "", golog.Lshortfile)
+	if secret == "" {
+		logger.Log("op", "startup", "warning", "no ml-secret-key set, memberlist traffic will not be encrypted")
+	} else {
+		sha := sha256.New()
+		mconfig.SecretKey = sha.Sum([]byte(secret))[:16]
+	}
+
+	// This channel is used by the Rejoin() method which runs on k8s node
+	// changes. A buffered channel is used to avoid blocking the main goroutine
+	// while waiting for the join operation to complete.
+	sl.mlJoinCh = make(chan struct{}, 1)
+
+	// ChannelEventDelegate hint that it should not block, so make mlEventCh 'big'
+	sl.mlEventCh = make(chan memberlist.NodeEvent, 1024)
+	mconfig.Events = &memberlist.ChannelEventDelegate{Ch: sl.mlEventCh}
+	var err error
+	sl.mList, err = memberlist.Create(mconfig)
+	if err != nil {
+		logger.Log("op", "startup", "error", err, "msg", "failed to create memberlist")
+		return nil, err
+	}
+	return &sl, nil
+}
+
+// Start starts the needed goroutines
+func (sl *SpeakerList) Start(client *k8s.Client) error {
+	sl.client = client
+
+	if sl.mList == nil {
+		return nil
+	}
+
+	go sl.memberlistWatchEvents()
+	go sl.mlJoinLoop()
+
+	return nil
+}
+
+// updateMlIPs runs forever updating the sl.mlSpeakersIPs slice with the
+// IPs of all the speaker pods. As this queries the API server, it waits at
+// least 5 minutes between queries, to avoid self-induced API server DoS (should
+// be trated with care).
+func (sl *SpeakerList) speakersIPsLoop() {
+	for {
+		// Create the ticker at the start of each loop, so we wait at
+		// least the ticker interval before the ticker case is true.
+		// If we create it outside of the loop, it will tick every 5
+		// minutes and if the blocking call happens to take longer, the
+		// next tick will be really soon and we may overload the API
+		// server.
+		ticker := time.NewTicker(5 * time.Minute)
+
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			// This blocks until the apiserver responds.
+			iplist, err := sl.mlSpeakers()
+			if err != nil {
+				sl.l.Log("op", "memberDiscovery", "error", err, "msg", "failed to get pod IPs")
+				continue
+			}
+
+			sl.mlMux.Lock()
+			sl.mlSpeakersIPs = iplist
+			sl.mlMux.Unlock()
+		}
+	}
+}
+
+func (sl *SpeakerList) mlSpeakers() ([]string, error) {
+	// This call blocks until we get a response of the API server.
+	// In the client-go version we are using, there is no way to use a
+	// context or timeout.
```

Comment:
can we change Config.Timeout 
https://godoc.org/k8s.io/client-go/rest#Config

Reply:
I think I checked this and it was removed in out client-go version, but I might have check it incorrectly (was kind of in a hurry :)). Will check again, thanks! :)

Reply:
Oh, I think I remember now. The clientset used here is shared across the code and stored in a struct. I think that was the reason to not modify that.

Thanks!